common/src/main/java/io/pravega/common/util/btree/BTreeIndex.java

```diff
@@ -0,0 +1,825 @@
+     * child pages, updating pointers to existing pages and inserting pointers for new pages (results of splits).
+     *
+     * @param parentPage The parent page.
+     * @param context    Processing Context.
+     */
+    private void processParentPage(PageWrapper parentPage, PageModificationContext context) {
+        if (context.getDeletedPageKey() != null) {
+            // We have a deleted page. Remove its pointer from the parent.
+            parentPage.getPage().delete(Collections.singleton(context.getDeletedPageKey()));
+            parentPage.markNeedsFirstKeyUpdate();
+        } else {
+            // Update parent page's child pointers for modified pages.
+            val toUpdate = context.getUpdatedPagePointers().stream()
+                                  .map(pp -> new PageEntry(pp.getKey(), serializePointer(pp)))
+                                  .collect(Collectors.toList());
+            parentPage.getPage().update(toUpdate);
+        }
+    }
+
+    /**
+     * Locates the Leaf Page that contains or should contain the given Key.
+     *
+     * @param key            A ByteArraySegment that represents the Key to look up the Leaf Page for.
+     * @param pageCollection A PageCollection that contains already looked-up pages. Any newly looked up pages will be
+     *                       inserted in this instance as well.
+     * @param timer          Timer for the operation.
+     * @return A CompletableFuture with a PageWrapper for the sought page.
+     */
+    private CompletableFuture<PageWrapper> locatePage(ByteArraySegment key, PageCollection pageCollection, TimeoutTimer timer) {
+        // Verify the sought key has the expected length.
+        Preconditions.checkArgument(key.getLength() == this.leafPageConfig.getKeyLength(), "Invalid key length.");
+
+        // Sanity check that our pageCollection has a somewhat correct view of the data index state. It is OK for it to
+        // think the index length is less than what it actually is (since a concurrent update may have increased it), but
+        // it is not a good sign if it thinks it's longer than the actual length.
+        Preconditions.checkArgument(pageCollection.getIndexLength() <= this.state.get().length, "Unexpected PageCollection.IndexLength.");
+
+        if (this.state.get().rootPageOffset == PagePointer.NO_OFFSET && pageCollection.getCount() == 0) {
+            // No data. Return an empty (leaf) page, which will serve as the root for now.
+            return CompletableFuture.completedFuture(pageCollection.insert(PageWrapper.wrapNew(createEmptyLeafPage(), null, null)));
+        }
+
+        AtomicReference<PagePointer> pagePointer = new AtomicReference<>(new PagePointer(null, this.state.get().rootPageOffset, this.state.get().rootPageLength));
+        CompletableFuture<PageWrapper> result = new CompletableFuture<>();
+        AtomicReference<PageWrapper> parentPage = new AtomicReference<>(null);
+        Futures.loop(
```

Comment:
Yes, the `loop` wasn't designed to return a non-void result. That's because, in general, Java loops do not return results, but execute a piece of code repeatedly until some condition happens. It does not know or care what that code does internally. For example, it may have an invariant complex enough that it will modify more than one variable, in which case it will be pretty hard to design a correct return value.